li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java

```diff
@@ -296,32 +300,31 @@ public void flush() {
 
   @Override
   public void close() {
-    LOG.info("Shutting down ...");
-    long start = System.currentTimeMillis();
-    prepClose();
-    _auditor.close();
-    _producer.close();
-    LOG.info("Shutdown complete in {} millis", (System.currentTimeMillis() - start));
+    close(Integer.MAX_VALUE, TimeUnit.MILLISECONDS);
   }
 
   @Override
   public void close(long timeout, TimeUnit timeUnit) {
-    LOG.info("Shutting down in {} {}...", timeout, timeUnit);
-    long start = System.currentTimeMillis();
-    long budget = timeUnit.toMillis(timeout);
-    long deadline = start + budget;
-    _auditor.close(budget, TimeUnit.MILLISECONDS);
-    long remaining = System.currentTimeMillis() - deadline; //could be negative
-    _producer.close(Math.max(0, remaining), TimeUnit.MILLISECONDS);
-    LOG.info("Shutdown complete in {} millis", (System.currentTimeMillis() - start));
-  }
+    LOG.info("Shutting down LiKafkaProducer in {} {}...", timeout, timeUnit);
+    long startTimeMs = System.currentTimeMillis();
+    long budgetMs = timeUnit.toMillis(timeout);
+    long deadlineTimeMs = startTimeMs + budgetMs;
 
-  private void prepClose() {
     _closed = true;
-    // We flush first to avoid the tight loop when buffer is full.
-    _producer.flush();
-    while (_numThreadsInSend.get() > 0) { }
-    _producer.flush();
+    while (_numThreadsInSend.get() > 0 && System.currentTimeMillis() < deadlineTimeMs) {
+      try {
+        synchronized (_numThreadsInSend) {
+          _numThreadsInSend.wait(deadlineTimeMs - System.currentTimeMillis());
```

Comment:
It seems we need to make sure the wait timeout is not negative. Otherwise an IllegalArgumentException may be thrown.

Reply:
@becketqin Thanks for catching this. I have fixed this issue and also updated the patch to fix a race condition issue.